File: test_dual_lane_tracking.m
function location = dual_lane_tracking_test(NumberOfObjects, u, reference)
% Test function to validate dual-lane center tracking logic
% This replicates the enhanced MATLAB function logic for validation

% Enhanced dual-lane center tracking implementation
% Detects left and right lanes separately and calculates center point

persistent last_left_lane last_right_lane last_center;

location = single([0 0]);

% Initialize persistent variables
if isempty(last_left_lane)
    last_left_lane = single([reference-100 0]);  % Offset left from reference
end
if isempty(last_right_lane)
    last_right_lane = single([reference+100 0]); % Offset right from reference  
end
if isempty(last_center)
    last_center = single([reference 0]);
end

% Process detected lane objects
if NumberOfObjects > 0
    % Convert input to matrix format for easier processing
    centroids = reshape(u, 2, [])';  % Each row is [x, y]
    
    if NumberOfObjects >= 2
        % Sort centroids by x-coordinate to identify left and right lanes
        [~, sort_idx] = sort(centroids(:,1));
        sorted_centroids = centroids(sort_idx, :);
        
        % Assign leftmost and rightmost as left and right lanes
        left_lane = sorted_centroids(1, :);
        right_lane = sorted_centroids(end, :);
        
        % Update persistent storage
        last_left_lane = single(left_lane);
        last_right_lane = single(right_lane);
        
        % Calculate center point between the two lanes
        center_x = (left_lane(1) + right_lane(1)) / 2;
        center_y = (left_lane(2) + right_lane(2)) / 2;
        last_center = single([center_x center_y]);
        
        fprintf('Dual lanes detected: Left=[%.1f,%.1f], Right=[%.1f,%.1f], Center=[%.1f,%.1f]\n', ...
                left_lane(1), left_lane(2), right_lane(1), right_lane(2), center_x, center_y);
        
    elseif NumberOfObjects == 1
        % Only one lane detected - determine if it's left or right
        detected_lane = centroids(1, :);
        
        % Compare with reference position to classify as left or right
        if detected_lane(1) < reference
            % Detected lane is on the left
            last_left_lane = single(detected_lane);
            % Estimate right lane position based on typical lane width (~200 pixels)
            estimated_right_x = detected_lane(1) + 200;
            % Calculate center point
            center_x = (detected_lane(1) + estimated_right_x) / 2;
            center_y = detected_lane(2);
            fprintf('Left lane detected: [%.1f,%.1f], Estimated right at %.1f, Center=[%.1f,%.1f]\n', ...
                    detected_lane(1), detected_lane(2), estimated_right_x, center_x, center_y);
        else
            % Detected lane is on the right  
            last_right_lane = single(detected_lane);
            % Estimate left lane position
            estimated_left_x = detected_lane(1) - 200;
            % Calculate center point
            center_x = (estimated_left_x + detected_lane(1)) / 2;
            center_y = detected_lane(2);
            fprintf('Right lane detected: [%.1f,%.1f], Estimated left at %.1f, Center=[%.1f,%.1f]\n', ...
                    detected_lane(1), detected_lane(2), estimated_left_x, center_x, center_y);
        end
        
        last_center = single([center_x center_y]);
    end
else
    % No lanes detected - use last known center position
    fprintf('No lanes detected, using last center: [%.1f,%.1f]\n', last_center(1), last_center(2));
end

% Output the calculated center location
location = last_center;

end

% Test cases
fprintf('=== Dual Lane Center Tracking Validation ===\n\n');

% Test 1: Dual lanes detected
fprintf('Test 1: Both lanes detected\n');
location1 = dual_lane_tracking_test(2, [300, 100, 500, 100], 400);

% Test 2: Only left lane detected  
fprintf('\nTest 2: Only left lane detected\n');
location2 = dual_lane_tracking_test(1, [300, 100], 400);

% Test 3: Only right lane detected
fprintf('\nTest 3: Only right lane detected\n');  
location3 = dual_lane_tracking_test(1, [500, 100], 400);

% Test 4: No lanes detected
fprintf('\nTest 4: No lanes detected\n');
location4 = dual_lane_tracking_test(0, [], 400);

fprintf('\n=== Test Complete ===\n');

File: test_dual_lane_tracking.py
#!/usr/bin/env python3
"""
Dual Lane Center Tracking Validation Script
Replicates the enhanced MATLAB function logic for testing
"""

import numpy as np

class DualLaneTracker:
    def __init__(self, reference=400):
        self.reference = reference
        self.last_left_lane = np.array([reference - 100, 0], dtype=np.float32)
        self.last_right_lane = np.array([reference + 100, 0], dtype=np.float32)
        self.last_center = np.array([reference, 0], dtype=np.float32)

    def process_lanes(self, num_objects, centroids):
        """
        Process detected lane objects and calculate center point
        
        Args:
            num_objects: Number of detected lane objects
            centroids: Flattened array of [x1, y1, x2, y2, ...] centroids
        
        Returns:
            center_location: [x, y] coordinates of calculated center
        """
        
        if num_objects > 0:
            # Convert flattened input to matrix format
            centroids_matrix = np.array(centroids).reshape(-1, 2)
            
            if num_objects >= 2:
                # Sort centroids by x-coordinate to identify left and right lanes
                sorted_indices = np.argsort(centroids_matrix[:, 0])
                sorted_centroids = centroids_matrix[sorted_indices]
                
                # Assign leftmost and rightmost as left and right lanes
                left_lane = sorted_centroids[0]
                right_lane = sorted_centroids[-1]
                
                # Update persistent storage
                self.last_left_lane = left_lane.astype(np.float32)
                self.last_right_lane = right_lane.astype(np.float32)
                
                # Calculate center point between the two lanes
                center_x = (left_lane[0] + right_lane[0]) / 2
                center_y = (left_lane[1] + right_lane[1]) / 2
                self.last_center = np.array([center_x, center_y], dtype=np.float32)
                
                print(f"Dual lanes detected: Left=[{left_lane[0]:.1f},{left_lane[1]:.1f}], "
                      f"Right=[{right_lane[0]:.1f},{right_lane[1]:.1f}], "
                      f"Center=[{center_x:.1f},{center_y:.1f}]")
                
            elif num_objects == 1:
                # Only one lane detected
                detected_lane = centroids_matrix[0]
                
                if detected_lane[0] < self.reference:
                    # Detected lane is on the left
                    self.last_left_lane = detected_lane.astype(np.float32)
                    # Estimate right lane position (typical lane width ~200 pixels)
                    estimated_right_x = detected_lane[0] + 200
                    center_x = (detected_lane[0] + estimated_right_x) / 2
                    center_y = detected_lane[1]
                    print(f"Left lane detected: [{detected_lane[0]:.1f},{detected_lane[1]:.1f}], "
                          f"Estimated right at {estimated_right_x:.1f}, "
                          f"Center=[{center_x:.1f},{center_y:.1f}]")
                else:
                    # Detected lane is on the right
                    self.last_right_lane = detected_lane.astype(np.float32)
                    # Estimate left lane position
                    estimated_left_x = detected_lane[0] - 200
                    center_x = (estimated_left_x + detected_lane[0]) / 2
                    center_y = detected_lane[1]
                    print(f"Right lane detected: [{detected_lane[0]:.1f},{detected_lane[1]:.1f}], "
                          f"Estimated left at {estimated_left_x:.1f}, "
                          f"Center=[{center_x:.1f},{center_y:.1f}]")
                
                self.last_center = np.array([center_x, center_y], dtype=np.float32)
        else:
            # No lanes detected - use last known center position
            print(f"No lanes detected, using last center: [{self.last_center[0]:.1f},{self.last_center[1]:.1f}]")
        
        return self.last_center

def run_tests():
    """Run validation tests for dual lane tracking"""
    print("=== Dual Lane Center Tracking Validation ===\n")
    
    tracker = DualLaneTracker(reference=400)
    
    # Test 1: Both lanes detected
    print("Test 1: Both lanes detected")
    location1 = tracker.process_lanes(2, [300, 100, 500, 100])
    
    # Test 2: Only left lane detected
    print("\nTest 2: Only left lane detected")
    location2 = tracker.process_lanes(1, [300, 100])
    
    # Test 3: Only right lane detected
    print("\nTest 3: Only right lane detected")
    location3 = tracker.process_lanes(1, [500, 100])
    
    # Test 4: Three lanes detected (should use leftmost and rightmost)
    print("\nTest 4: Three lanes detected")
    location4 = tracker.process_lanes(3, [250, 100, 400, 105, 550, 100])
    
    # Test 5: No lanes detected
    print("\nTest 5: No lanes detected")
    location5 = tracker.process_lanes(0, [])
    
    print("\n=== Test Complete ===")
    
    # Verify expected behavior
    print("\n=== Verification ===")
    print(f"Final center location: [{location5[0]:.1f}, {location5[1]:.1f}]")
    print("✓ All test cases completed successfully")

if __name__ == "__main__":
    run_tests()